cmd/influxd/inspect/report_tsi/report_tsi.go

```diff
@@ -0,0 +1,481 @@
+		c, ok := report.cardinalities[shardID][string(name)]
+		if !ok {
+			c = &cardinality{name: name}
+			report.cardinalities[shardID][string(name)] = c
+		}
+
+		sitr, err := idx.MeasurementSeriesIDIterator(name)
+		if err != nil {
+			return err
+		} else if sitr == nil {
+			continue
+		}
+
+		var e tsdb.SeriesIDElem
+		for e, err = sitr.Next(); err == nil && e.SeriesID != 0; e, err = sitr.Next() {
+			if e.SeriesID > math.MaxUint32 {
+				panic(fmt.Sprintf("series ID is too large: %d (max %d). Corrupted series file?", e.SeriesID, uint32(math.MaxUint32)))
+			}
+			c.add(e.SeriesID)
+		}
+		sitr.Close()
+
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type result struct {
+	name  []byte
+	count int64
+
+	// For low cardinality measurements just track series using map
+	lowCardinality map[uint32]struct{}
+
+	// For higher cardinality measurements track using bitmap.
+	set *tsdb.SeriesIDSet
+}
+
+func (r *result) addShort(ids []uint32) {
+	// There is already a bitset of this result.
+	if r.set != nil {
+		for _, id := range ids {
+			r.set.AddNoLock(uint64(id))
+		}
+		return
+	}
+
+	// Still tracking low cardinality sets
+	if r.lowCardinality == nil {
+		r.lowCardinality = map[uint32]struct{}{}
+	}
+
+	for _, id := range ids {
+		r.lowCardinality[id] = struct{}{}
+	}
+
+	// Cardinality is large enough that we will benefit from using a bitmap
+	if len(r.lowCardinality) > useBitmapN {
+		r.set = tsdb.NewSeriesIDSet()
+		for id := range r.lowCardinality {
+			r.set.AddNoLock(uint64(id))
+		}
+		r.lowCardinality = nil
+	}
+}
+
+func (r *result) merge(other *tsdb.SeriesIDSet) {
+	if r.set == nil {
+		r.set = tsdb.NewSeriesIDSet()
+		for id := range r.lowCardinality {
+			r.set.AddNoLock(uint64(id))
+		}
+		r.lowCardinality = nil
+	}
+	r.set.Merge(other)
+}
+
+type results []*result
+
+func (a results) Len() int           { return len(a) }
+func (a results) Less(i, j int) bool { return a[i].count < a[j].count }
+func (a results) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+
+func (report *reportTSI) printSummaryByMeasurement() error {
+	// Get global set of measurement names across shards.
+	idxs := &tsdb.IndexSet{SeriesFile: report.sfile}
+	for _, idx := range report.shardIdxs {
+		idxs.Indexes = append(idxs.Indexes, idx)
+	}
+
+	mitr, err := idxs.MeasurementIterator()
+	if err != nil {
+		return err
+	} else if mitr == nil {
+		return errors.New("got nil measurement iterator for index set")
+	}
+	defer mitr.Close()
+
+	var name []byte
+	var totalCardinality int64
+	measurements := results{}
+	for name, err = mitr.Next(); err == nil && name != nil; name, err = mitr.Next() {
+		res := &result{name: name}
+		for _, shardCards := range report.cardinalities {
+			other, ok := shardCards[string(name)]
+			if !ok {
+				continue // this shard doesn't have anything for this measurement.
+			}
+
+			if other.short != nil && other.set != nil {
+				panic("cardinality stored incorrectly")
```

Comment:
I'm guessing this `panic` is here because this case should never happen. In that case I'd say we drop the check entirely.